reaper-adm-extension/src/reaper_adm/update_check.cpp

```diff
@@ -0,0 +1,210 @@
+#include "update_check.h"
+#include <version/eps_version.h>
+#include <iostream>
+#include <vector>
+#include <helper/resource_paths_juce-file.hpp>
+#include <helper/native_message_box.hpp>
+
+UpdateChecker::UpdateChecker()
+{
+    if (eps::versionInfoAvailable()) {
+        currentVersion = Version(
+            eps::versionMajor(),
+            eps::versionMinor(),
+            eps::versionRevision());
+    }
+
+    settingsFile = ResourcePaths::getSettingsDirectory(true).getChildFile("UpdateCheck.settings");
+    if (!settingsFileExists()) {
+        // No settings file - perhaps first run?
+        if (saveSettings() && settingsFileExists()) {
+            // Settings file is writable so probably was just first run.
+            // Set autocheck on initially.
+            setAutoCheckEnabled(true);
+        }
+    }
+    loadSettings();
+}
+
+UpdateChecker::~UpdateChecker()
+{
+}
+
+bool UpdateChecker::getAutoCheckEnabled()
+{
+    return settingAutoCheckEnabled;
+}
+
+bool UpdateChecker::setAutoCheckEnabled(bool enabled, bool displayConfirmation)
+{
+    settingAutoCheckEnabled = enabled;
+    auto success = saveSettings();
+    if (displayConfirmation) {
+        if (success) {
+            if (enabled) {
+                displayMessageBox(messageBoxTitles, "Update check will now be performed each time REAPER is started.", MB_ICONINFORMATION);
+            }
+            else {
+                displayMessageBox(messageBoxTitles, "Update check will no longer be performed when REAPER is started.", MB_ICONINFORMATION);
+            }
+        }
+        else {
+            displayMessageBox(messageBoxTitles, "Error: Failed to save preferences.", MB_ICONEXCLAMATION);
+        }
+    }
+    return success;
+}
+
+void UpdateChecker::doUpdateCheck(bool manualCheck, int timeoutMs)
+{
+    std::string body;
+    auto getSuccess = getHTTPResponseBody(versionJsonUrl, body, timeoutMs);
+    if (!getSuccess) {
+        if (manualCheck) {
+            displayError("Failed to get data.");
+        }
+        return;
+    }
+
+    juce::var j;
+    auto parseResult = juce::JSON::parse(body, j);
+    if (parseResult.failed()) {
+        if (manualCheck) {
+            displayError("Failed to parse data.");
+        }
+        return;
+    }
+
+    juce::var varVersionMajor = j.getProperty("version_major", juce::var());
+    juce::var varVersionMinor = j.getProperty("version_minor", juce::var());
+    juce::var varVersionRevision = j.getProperty("version_revision", juce::var());
+
+    if (!varVersionMajor.isInt() ||
+        !varVersionMinor.isInt() ||
+        !varVersionRevision.isInt()) {
+        if (manualCheck) {
+            displayError("Unexpected data.");
+        }
+        return;
+    }
+
+    Version remoteVersion(

```

Comment:
I might split out everything up to here into a `std::optional<Version> get_remote_version(bool manualCheck, int timeoutMs)` function, as fetching/parsing and processing the result feel like two quite distinct steps.